### PR TITLE
azurerm_api_management_api - support for `subscription_required`

### PR DESCRIPTION
**Pull request by brunhil**
Add `subscription_required` to data and resource for **azurerm_api_management_api**

(fixes ##3863 )

### CODE REVIEW

File: azurerm/resource_arm_api_management_api.go [hunk 1]
```diff
@@ -454,3 +477,30 @@ func flattenApiManagementApiSubscriptionKeyParamNames(paramNames *apimanagement.
...
 	return []interface{}{result}
 }
+
+func expandApiManagementAuthenticationSettings(input []interface{}) *apimanagement.AuthenticationSettingsContract {
+	if len(input) == 0 {
+		return nil
+	}
+
+	v := input[0].(map[string]interface{})
+	subscriptionKeyRequired := v["subscriptionKeyRequired"].(bool)
```
Comment (mbfrahry) [inline azurerm/resource_arm_api_management_api.go]:
```suggestion
	subscriptionKeyRequired := v["subscription_key_required"].(bool)
```

File: website/docs/d/api_management_api.html.markdown [hunk 2]
```diff
@@ -58,12 +58,20 @@ output "api_management_api_id" {
...
+
 * `version` - The Version number of this API, if this API is versioned.
 
 * `version_set_id` - The ID of the Version Set which this API is associated with.
 
 ---
 
+An `authentication_settings` block exports the following:
+
+* `subscription_key_required` - Should this API require a subscription key?
```
Comment (mbfrahry) [inline website/docs/d/api_management_api.html.markdown]:
```suggestion
* `subscription_key_required` - Whether the API requires a subscription key or not.
```

**Comment by brunhil** [comment]
Good catch with the spelling @mbfrahry , it was one of those days yesterday -_-. I'll get the tests added and other changes fixed and push them up.

File: azurerm/resource_arm_api_management_api_test.go [hunk 3]
```diff
@@ -28,6 +28,8 @@ func TestAccAzureRMApiManagementApi_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "soap_pass_through", "false"),
 					resource.TestCheckResourceAttr(resourceName, "is_current", "true"),
 					resource.TestCheckResourceAttr(resourceName, "is_online", "false"),
+					resource.TestCheckResourceAttr(resourceName, "authentication_settings.#", "1"),
```
Comment (brunhil) [inline azurerm/resource_arm_api_management_api_test.go]: This is set by default on creation of APIs for APIM, so I thought it was easiest to verify on the "basic" test. If you want an explicit test let me know.

File: azurerm/resource_arm_api_management_api.go [hunk 4]
```diff
@@ -454,3 +477,30 @@ func flattenApiManagementApiSubscriptionKeyParamNames(paramNames *apimanagement.
 
 	return []interface{}{result}
 }
+
+func expandApiManagementAuthenticationSettings(input []interface{}) *apimanagement.AuthenticationSettingsContract {
+	if len(input) == 0 {
+		return nil
+	}
+
+	v := input[0].(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_api_management_api.go]: This caused a test to panic so we should confirm that `input[0] != nil` before casting this to a `map[string]interface{}`

```

------- Stdout: -------
=== RUN   TestAccAzureRMApiManagementApi_importUpdate

------- Stderr: -------
2019/11/15 17:21:04 [DEBUG] Registering Data Sources for "Compute"..
2019/11/15 17:21:04 [DEBUG] Registering Resources for "Compute"..
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1247 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm.expandApiManagementAuthenticationSettings(...)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_api_management_api.go:486
github.com/terraform-providers/terraform-provider-azurerm/azurerm.resourceArmApiManagementApiCreateUpdate(0xc0002c9490, 0x3b50f20, 0xc0014fe000, 0x0, 0x0)
```

File: azurerm/resource_arm_api_management_api_test.go [hunk 5]
```diff
@@ -28,6 +28,8 @@ func TestAccAzureRMApiManagementApi_basic(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "authentication_settings.0.subscription_key_required", "true"),
```
Comment (mbfrahry) [inline azurerm/resource_arm_api_management_api_test.go]: This test now errors. It doesn't look like this is being set properly
```
------- Stdout: -------
=== RUN   TestAccAzureRMApiManagementApi_basic
--- FAIL: TestAccAzureRMApiManagementApi_basic (2612.17s)
    testing.go:569: Step 0 error: Check failed: Check 6/6 error: azurerm_api_management_api.test: Attribute 'authentication_settings.0.subscription_key_required' not found
FAIL
```

File: azurerm/resource_arm_api_management_api_test.go [hunk 6]
```diff
@@ -120,6 +122,33 @@ func TestAccAzureRMApiManagementApi_version(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMApiManagementApi_authenticationSettings(t *testing.T) {
```
Comment (mbfrahry) [inline azurerm/resource_arm_api_management_api_test.go]: This test is also failing

```
=== RUN   TestAccAzureRMApiManagementApi_authenticationSettings
--- FAIL: TestAccAzureRMApiManagementApi_authenticationSettings (2825.43s)
    testing.go:569: Step 0 error: Check failed: Check 3/3 error: azurerm_api_management_api.test: Attribute 'authentication_settings.0.subscription_key_required' not found
FAIL
```

File: azurerm/resource_arm_api_management_api.go [hunk 7]
```diff
@@ -159,6 +159,21 @@ func resourceArmApiManagementApi() *schema.Resource {
 				},
 			},
 
+			"authentication_settings": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
```
[the remark this answers is not in the document]
Comment (brunhil) [reply azurerm/resource_arm_api_management_api.go]: So this actually is a misunderstanding on my part...I'll explain more in the message on this PR.

**Comment by brunhil** [comment]
After digging further into this, it turns out we have a few approaches we can take with this.

The 'AuthenticationSettings' field is a 2019-01-01 API Version property. So these changes are actually for 2019-01-01. I was implementing this when looking at the REST API [here](https://docs.microsoft.com/en-us/rest/api/apimanagement/2019-01-01/apis/createorupdate). I had modified the try-it option to point at 2018-01-1 and assumed the documentation on the left updated, it did not :).

After testing/verifying in 2018-01-01, the response body is as follows:

```
{
  "id": "...",
  "type": "Microsoft.ApiManagement/service/apis",
  "name": "test-api-name",
  "properties": {
    "displayName": "test-api-name",
    "apiRevision": "1",
    "description": "",
    "subscriptionRequired": false,
```

Then attempting to set the value in the body for the REST API led to,

```
{
  "error": {
    "code": "ValidationError",
    "message": "SubscriptionRequired property is not supported for versions before 20180601.",
    "details": null
  }
}
```

As a result, I'm marking this PR as WIP. We would need to update to 2019-01-01 API version to support this field, and this is blocked by the upstream issues that #2613 is blocked by.

**Comment by tombuildsstuff** [comment]
Assigning this to the "blocked" milestone since this is blocked on upgrading to 2019-01-01, which is blocked on https://github.com/Azure/azure-rest-api-specs/issues/6372

**Comment by katbyte** [comment]
Looks like this should be close from the blocking issue
```
just yesterday we got signed all required PRs and now preparing the nuget package. Should be ready in nearest days.
``

**Comment by usamaahmadkhan** [comment]
@brunhil can we have an estimated date when this will get released? Blocked as we want to consume this resource from terraform in our module

**Comment by markti** [comment]
what about linking to an authorization_server (oauth / openidc) ??? is this included in this PR?

**Comment by katbyte** [comment]
@brunhil - apim api as been updated on master so this should be unblocked now

**Comment by brunhil** [comment]
Thanks for the note @katbyte , I'll start getting this up to date!

**Comment by markti** [comment]
this should resolve issue #3863 no?

**Comment by brunhil** [comment]
@markti - This PR was simply to expose the ability to turn off Subscription Settings, so that won't be added

**Comment by markti** [comment]
> @markti - This PR was simply to expose the ability to turn off Subscription Settings, so that won't be added

Ok, no problem. This is a great submission as its very needed functionality. 

I'd like to implement the Oauth / OIDC setting...very eager to see this merged so I can work on that...

**Comment by brunhil** [comment]
@markti - It turns out the AuthenticationSettings block isn't required any more for this PR as the **SubscriptionRequired** block is now at the top level.

**Comment by brunhil** [comment]
@katbyte , when this was initially implemented Subscription Required was turned on by default. They may have changed their behavior in the new API version.

**Comment by danderton** [comment]
We're also blocked with this current issue. Do you have an ETA?

**Comment by brunhil** [comment]
Sorry all, forgot to hit push! The test should be fixed if it defaults to false. Let me know if there's any other requested changes. Thanks for reviewing!

**Comment by 3mard** [comment]
Thanks you all for this, can't wait for this to get merged

**Pull request merged**